### PR TITLE
wallets restructuring #272

### PR DESCRIPTION
**Pull request by adecaro**
Signed-off-by: Angelo De Caro <adc@zurich.ibm.com>

### CODE REVIEW

**Review by mbrandenburger — COMMENTED** [review]
Huge thing :)

File: token/core/identity/provider.go [hunk 1]
```diff
@@ -13,144 +14,100 @@ import (
...
 	}
 }
 
-func (i *Provider) GetIdentityInfo(usage driver.IdentityUsage, id string) *driver.IdentityInfo {
-	mapper, ok := i.mappers[usage]
+func (p *Provider) GetIdentityInfo(usage driver.IdentityRole, id string) driver.IdentityInfo {
+	wallet, ok := p.wallets[usage]
 	if !ok {
-		panic(fmt.Sprintf("mapper not found for usage [%d]", usage))
+		panic(fmt.Sprintf("wallet not found for usage [%d]", usage))
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/core/identity/provider.go:61]: @mbrandenburger , I get rid of the panics in this PR or another one. What do you think?

File: token/core/identity/msp/idemix/lm.go [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+
+func (lm *LocalMembership) IsMe(id view.Identity) bool {
+	return lm.signerService.IsMe(id)
+}
+
+func (lm *LocalMembership) GetIdentifier(id view.Identity) (string, error) {
+	label := string(id)
+	if logger.IsEnabledFor(zapcore.DebugLevel) {
+		logger.Debugf("get anonymous identity info by label [%s]", label)
+	}
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/core/identity/msp/idemix/lm.go:114]: can I address this in another dedicated PR?

File: integration/nwo/token/generators/generators.go [hunk 3]
```diff
@@ -7,11 +7,24 @@ SPDX-License-Identifier: Apache-2.0
 package generators
 
 import (
+	api2 "github.com/hyperledger-labs/fabric-smart-client/integration/nwo/api"
+	"github.com/hyperledger-labs/fabric-smart-client/integration/nwo/common"
 	"github.com/hyperledger-labs/fabric-smart-client/integration/nwo/fsc/node"
-
 	"github.com/hyperledger-labs/fabric-token-sdk/integration/nwo/token/topology"
+	"github.com/onsi/gomega/gexec"
 )
```
Comment (KElkhiyaoui) [inline integration/nwo/token/generators/generators.go:16]: shall we add comments here to describe the methods?

File: token/core/identity/msp/idemix/wallet.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+type localMembership interface {
+	DefaultNetworkIdentity() view.Identity
+	IsMe(id view.Identity) bool
+	GetIdentityInfo(label string, auditInfo []byte) (driver.IdentityInfo, error)
+	GetIdentifier(id view.Identity) (string, error)
+	GetDefaultIdentifier() string
+	RegisterIdentity(id string, path string) error
+}
+
+// wallet maps identifiers of different sorts to identities
```
Comment (KElkhiyaoui) [inline token/core/identity/msp/idemix/wallet.go:26]: wallet maps an identifier to an identity.

File: token/core/identity/msp/x509/mspx509.go [hunk 5]
```diff
@@ -9,14 +9,16 @@ package x509
...
 
+	"github.com/IBM/idemix/common/flogging"
 	"github.com/golang/protobuf/proto"
 	"github.com/hyperledger-labs/fabric-smart-client/platform/view/view"
+	"github.com/hyperledger-labs/fabric-token-sdk/token/driver"
 	"github.com/hyperledger/fabric-protos-go/msp"
 	"github.com/pkg/errors"
-
-	"github.com/hyperledger-labs/fabric-token-sdk/token/driver"
 )
```
Comment (KElkhiyaoui) [inline token/core/identity/msp/x509/mspx509.go:19]: token-sdk.core.x509?

File: token/core/identity/msp/x509/wallet.go [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+type localMembership interface {
+	DefaultNetworkIdentity() view.Identity
+	IsMe(id view.Identity) bool
+	GetIdentityInfo(label string, auditInfo []byte) (driver.IdentityInfo, error)
+	GetIdentifier(id view.Identity) (string, error)
+	GetDefaultIdentifier() string
+	RegisterIdentity(id string, path string) error
+}
+
```
Comment (KElkhiyaoui) [inline token/core/identity/msp/x509/wallet.go:25]: maps an identifier to an identity.

File: token/core/identity/msp/x509/wallet.go [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+		nodeIdentity:    nodeIdentity,
+		localMembership: localMembership,
+	}
+}
+
+// GetIdentityInfo get in input an identifier and returns:
+// - The corresponding long term identifier
+// - The corresponding enrollment ID
+// - A function that returns the identity and its audit info.
+func (i *wallet) GetIdentityInfo(id string) driver.IdentityInfo {
```
Comment (KElkhiyaoui) [inline token/core/identity/msp/x509/wallet.go:45]: change receiver to w?

File: token/core/identity/msp/wm.go [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+	CertifierMSPID = "CertifierMSPID"
+)
+
+// IdentityType is the type of identity
+type IdentityType int
+
+const (
+	// LongTermIdentity is the type of the long term identity (x509)
+	LongTermIdentity IdentityType = iota
+	// AnonymousIdentity is the type of the anonymous identity (Idemix
```
Comment (KElkhiyaoui) [inline token/core/identity/msp/wm.go:41]: mention that this is also a long term identity

File: token/core/identity/provider.go [hunk 9]
```diff
@@ -272,8 +223,44 @@ func (i *Provider) Bind(id view.Identity, to view.Identity) error {
...
 	}
 	return nil
 }
+
+func (p *Provider) AddDeserializer(d Deserializer) {
+	p.deserializers = append(p.deserializers, d)
+}
+
+// Info wraps a driver.IdentityInfo to further register the audit info,
+// and bind the new identity to the default FSC node identity
```
Comment (KElkhiyaoui) [inline token/core/identity/provider.go:237]: and binds

File: token/core/identity/wallet.go [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+// Wallets is a map of Wallet, one for each identity role
+type Wallets map[driver.IdentityRole]Wallet
+
+// NewWallets returns a new Wallets maps
+func NewWallets() Wallets {
+	return make(Wallets)
+}
+
+// Put associated a wallet to a given identity role
```
Comment (KElkhiyaoui) [inline token/core/identity/wallet.go:32]: associates

File: token/driver/identity.go [hunk 11]
```diff
@@ -3,49 +3,61 @@ Copyright IBM Corp. All Rights Reserved.
...
 	CertifierRole
 )
 
-type GetIdentityFunc func() (view.Identity, error)
-
-type IdentityInfo struct {
-	ID           string
-	EnrollmentID string
-	GetIdentity  GetIdentityFunc
+// IdentityInfo models and Identity inside the Identity Provider
```
Comment (KElkhiyaoui) [inline token/driver/identity.go:27]: an identity?

File: token/wallet.go [hunk 12]
```diff
@@ -40,8 +40,8 @@ func (t *WalletManager) IsMe(id view.Identity) bool {
 }
 
 // RegisterOwnerWallet registers a new owner wallet with type passed id
-func (t *WalletManager) RegisterOwnerWallet(id string, typ string, path string) error {
-	return t.managementService.tms.RegisterOwnerWallet(id, typ, path)
+func (t *WalletManager) RegisterOwnerWallet(id string, path string) error {
```
Comment (KElkhiyaoui) [inline token/wallet.go:43]: change receiver to wm?

**Review by KElkhiyaoui — CHANGES_REQUESTED** [review]

File: token/wallet.go (hunk 12 above)
Comment (adecaro) [reply token/wallet.go:43]: okay

File: token/driver/identity.go (hunk 11 above)
Comment (adecaro) [reply token/driver/identity.go:27]: okay

File: token/core/identity/wallet.go (hunk 10 above)
Comment (adecaro) [reply token/core/identity/wallet.go:32]: okay

File: token/core/identity/provider.go (hunk 9 above)
Comment (adecaro) [reply token/core/identity/provider.go:237]: okay

File: token/core/identity/msp/wm.go (hunk 8 above)
Comment (adecaro) [reply token/core/identity/msp/wm.go:41]: okay

File: token/core/identity/msp/x509/wallet.go (hunk 6 above)
Comment (adecaro) [reply token/core/identity/msp/x509/wallet.go:25]: okay

File: token/core/identity/msp/x509/wallet.go (hunk 7 above)
Comment (adecaro) [reply token/core/identity/msp/x509/wallet.go:45]: okay

File: token/core/identity/msp/idemix/wallet.go (hunk 4 above)
Comment (adecaro) [reply token/core/identity/msp/idemix/wallet.go:26]: okay

File: token/core/identity/msp/idemix/wallet.go [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+func NewWallet(networkID string, nodeIdentity view.Identity, localMembership localMembership) *wallet {
+	return &wallet{
+		networkID:       networkID,
+		nodeIdentity:    nodeIdentity,
+		localMembership: localMembership,
+	}
+}
+
+// GetIdentityInfo get in input an identifier and returns:
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/core/identity/msp/idemix/wallet.go:41]: oka

File: token/core/identity/msp/idemix/wallet.go [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+// - The corresponding long term identifier
+// - The corresponding enrollment ID
+// - A function that returns the identity and its audit info.
+func (i *wallet) GetIdentityInfo(id string) driver.IdentityInfo {
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/core/identity/msp/idemix/wallet.go:45]: okay

File: token/core/identity/msp/idemix/wallet.go [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+	info, err := i.localMembership.GetIdentityInfo(id, nil)
+	if err != nil {
+		if logger.IsEnabledFor(zapcore.DebugLevel) {
+			logger.Debugf("[%s] failed to get anonymous identity for [%s]: %s", i.networkID, id, err)
+		}
+		return nil
+	}
+	return info
+}
+
```
[the remark this answers is not in the document]
Comment (adecaro) [reply token/core/identity/msp/idemix/wallet.go:59]: okay

File: integration/nwo/token/generators/generators.go (hunk 3 above)
Comment (adecaro) [reply integration/nwo/token/generators/generators.go:16]: okay

File: token/core/identity/msp/idemix/lm.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (mbrandenburger) [reply token/core/identity/msp/idemix/lm.go:114]: if you prefer ... :)

[the remark this answers is not in the document]
Comment (adecaro) [reply token/core/identity/msp/idemix/lm.go:114]: perfect

File: integration/nwo/token/generators/generators.go [hunk 16]
```diff
@@ -14,25 +14,43 @@ import (
...
-	PublicParametersFile(tms *topology.TMS) string
-	GetContext() api2.Context
-	PublicParameters(tms *topology.TMS) []byte
+	// GetPublicParamsGenerators returns the public parameters' generator for the given driver
 	GetPublicParamsGenerators(driver string) PublicParamsGenerator
+	// PublicParameters returns the public parameters of the given TMS
+	PublicParameters(tms *topology.TMS) []byte
+	// PublicParametersFile returns the path to the public parameters file of the given TMS
+	PublicParametersFile(tms *topology.TMS) string
+	// PublicParametersDir return the path to the directory containing the public parameters
```
Comment (KElkhiyaoui) [inline integration/nwo/token/generators/generators.go:27]: returns

Comment (adecaro) [reply integration/nwo/token/generators/generators.go]: fixed

File: Makefile [hunk 17]
```diff
@@ -52,8 +52,8 @@ orion-server-images:
 .PHONY: dependencies
 dependencies:
 	go install github.com/onsi/ginkgo/ginkgo
-	go install github.com/gordonklaus/ineffassign
-	go install github.com/google/addlicense
+	go install github.com/gordonklaus/ineffassign@4cc7213b9bc8b868b2990c372f6fa057fa88b91c
+	go install github.com/google/addlicense@2fe3ee94479d08be985a84861de4e6b06a1c7208
```
Comment (mbrandenburger) [inline Makefile:56]: do you still want to include this?

Comment (adecaro) [reply Makefile:56]: nope, let me clean

**Review by KElkhiyaoui — APPROVED** [review]

**Comment by mbrandenburger** [comment]
@KElkhiyaoui sorry, clicked the wrong button and thereby converted your approval into another review request.

**Review by mbrandenburger — APPROVED** [review]
LGTM

**Review by KElkhiyaoui — APPROVED** [review]

**Pull request merged**